tests/integration/models/reddit/test_rules.py

```
@@ -0,0 +1,139 @@
+    def test_delete_rule(self, _):
+        self.reddit.read_only = False
+        with self.recorder.use_cassette("TestRule.test_delete_rule"):
+            rule = self.subreddit.rules[-1]
+            rule.delete()
+
+    def test_get_rules(self):
+        with self.recorder.use_cassette("TestRule.test_get_rules"):
+            for rule in self.subreddit.rules:
+                assert isinstance(rule, Rule)
+
+    def test_get_rule_string(self):
+        with self.recorder.use_cassette("TestRule.test_get_rule_string"):
+            rule = self.subreddit.rules["PRAW Test"]
+            assert isinstance(rule, Rule)
+            assert rule.short_name == "PRAW Test"
+
+    def test_get_rule_int(self):
+        with self.recorder.use_cassette("TestRule.test_get_rule_int"):
+            assert isinstance(self.subreddit.rules[0], Rule)
+
+    def test_get_rule_negative_int(self):
+        with self.recorder.use_cassette("TestRule.test_get_rule_int"):
+            assert isinstance(self.subreddit.rules[-1], Rule)
+
+    @mock.patch("time.sleep", return_value=None)
+    def test_reorder_rules(self, _):
+        self.reddit.read_only = False
+        with self.recorder.use_cassette("TestRule.test_reorder_rules"):
+            rule_list = list(self.subreddit.rules)
+            reordered = rule_list[2:3] + rule_list[0:2] + rule_list[3:]
+            rule_info = {rule.short_name: rule for rule in rule_list}
+            self.subreddit.rules.reorder(reordered)
+            for rule in self.subreddit.rules:
+                assert rule_info[rule.short_name] == rule
+
+    @mock.patch("time.sleep", return_value=None)
+    def test_reorder_rules_no_reorder(self, _):
+        self.reddit.read_only = False
+        with self.recorder.use_cassette(
+            "TestRule.test_reorder_rules_no_reorder"
+        ):
+            rule_list = list(self.subreddit.rules)
+            assert self.subreddit.rules.reorder(rule_list) == rule_list
+
```

Could you also add a test where you try to set an invalid reordering? Types of invalidity I can think of are:

- empty list
- some rules are missing from the list
- some rules appear twice in the list
- some non-rule is in the list

---

I would assume that the fourth method would fail because it's missing the attribute `short_name`, but the other combinations could surely be tried.

---

Right — for the cases that fail, assert that a certain type of exception is raised, and maybe even assert something about that exception. We want exceptional circumstances to throw exceptions and not just silently fail.

---

Or for case 4, try instead to use a rule that has a shortname that Reddit doesn't know about.